src/btree/bt_split.c

```diff
@@ -1220,21 +1233,55 @@ __split_internal_lock(WT_SESSION_IMPL *session, WT_REF *ref, bool trylock, WT_PA
         WT_RET(__wt_page_modify_init(session, parent));
 
         if (trylock) {
-            WT_RET(WT_PAGE_TRYLOCK(session, parent));
-        } else
+            /*
+             * If the parent's children have all be evicted, the parent itself can be evicted after
+             * the new page index is visible and before the split generation is set. Acquire a
+             * hazard pointer to prevent the parent being evicted before we set the page split
+             * generation.
+             */
+#ifdef HAVE_DIAGNOSTIC
+            WT_RET(__wt_hazard_set_func(
+              session, parent->pg_intl_parent_ref, &busy, __PRETTY_FUNCTION__, __LINE__));
+#else
+            WT_RET(__wt_hazard_set_func(session, parent->pg_intl_parent_ref, &busy));
+#endif
+            if (busy)
+                return (EBUSY);
+            ret = WT_PAGE_TRYLOCK(session, parent);
+            if (ret != 0) {
+                WT_TRET(__wt_hazard_clear(session, parent->pg_intl_parent_ref));
+                WT_RET(ret);
+            }
+        } else {
+            /*
+             * If the parent's children have all be evicted, the parent itself can be evicted after
+             * the new page index is visible and before the split generation is set. Acquire a
+             * hazard pointer to prevent the parent being evicted before we set the page split
+             * generation.
+             */
+            do {
+#ifdef HAVE_DIAGNOSTIC
+                WT_RET(__wt_hazard_set_func(
+                  session, parent->pg_intl_parent_ref, &busy, __PRETTY_FUNCTION__, __LINE__));
+#else
+                WT_RET(__wt_hazard_set_func(session, parent->pg_intl_parent_ref, &busy));
+#endif
+                if (busy)
+                    __wt_yield();
+            } while (busy);
             WT_PAGE_LOCK(session, parent);
+        }
         if (parent == ref->home)
             break;
         WT_PAGE_UNLOCK(session, parent);
+        WT_RET(__wt_hazard_clear(session, parent->pg_intl_parent_ref));
     }
 
     /*
      * This child has exclusive access to split its parent and the child's existence prevents the
      * parent from being evicted. However, once we update the parent's index, it may no longer refer
-     * to the child, and could conceivably be evicted. If the parent page is dirty, our page lock
-     * prevents eviction because reconciliation is blocked. However, if the page were clean, it
-     * could be evicted without encountering our page lock. That isn't possible because you cannot
-     * move a child page and still leave the parent page clean.
+     * to the child, and could conceivably be evicted. Therefore, we need to acquire a hazard
+     * pointer and the page lock to prevent it from being evicted.
```

Comment:
I think it is now redundant to take the page lock. I will try separately to not take the page lock. But it is risky.

Reply:
Here's the [patch](https://spruce.mongodb.com/version/68120c26775f9e000753fbfa/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC).

Reply:
It doesn't work as we still need the lock to prevent to split happen on the same page.